### PR TITLE
[FLINK-7251] [types] Remove the flink-java8 module and improve lambda type extraction

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR merge the `flink-java8` module into `flink-core`/`flink-runtime`. It also fixes issues with lambda extraction that occurred outside of `flink-java8` which used a different compiler.

Lambdas are still difficult to use in Flink. I'm working on an experimental solution that I will publish soon.

## Brief change log

- Integration of some lambda functions into existing examples/tests
- Removal of error-prone lambda input validation
- Removal of `flink-java8`
- Code clean up (remove warnings at several places)

## Verifying this change

This change is already covered by existing tests, such as `LambdaExtractionTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by twalthr** [comment]
@zentol are there other locations where I have to remove the `flink-java8` module?

**Comment by twalthr** [comment]
I checked the docs. They do not change with these changes. However, we should verify that the docs about the Eclipse JDT compiler are still correct maybe even with an end-to-end test.

**Comment by twalthr** [comment]
@StephanEwen could you take a look?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 1]
```diff
@@ -584,21 +581,6 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
 				// number of parameters the SAM of implemented interface has; the parameter indexing applies to this range
 				final int baseParametersLen = sam.getParameterTypes().length;
 
-				// executable references "this" implicitly
-				if (paramLen <= 0) {
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Why you removed checking the input type information?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 2]
```diff
@@ -584,21 +581,6 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
-					// executable declaring class can also be a super class of the input type
-					// we only validate if the executable exists in input type
-					validateInputContainsExecutable(exec, inType);
-				}
-				else {
-					final Type input = TypeExtractionUtils.extractTypeFromLambda(
-						exec,
-						lambdaInputTypeArgumentIndices,
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: If you removed checking the input type info the parameter `lambdaInputTypeArgumentIndices` is no longer necessary.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 3]
```diff
@@ -695,29 +678,14 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
-				// check for lambda type erasure
-				validateLambdaGenericParameters(exec);
 
 				final Method sam = TypeExtractionUtils.getSingleAbstractMethod(baseClass);
 				final int baseParametersLen = sam.getParameterTypes().length;
 
 				// parameters must be accessed from behind, since JVM can add additional parameters e.g. when using local variables inside lambda function
 				final int paramLen = exec.getParameterTypes().length;
 
-				final Type input1 = TypeExtractionUtils.extractTypeFromLambda(
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Same comments as above. Why don't check the input types, if not checking them the lambda input indices not necessary.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 4]
```diff
@@ -119,19 +120,15 @@ public boolean filter(Event value) throws Exception {
 			}
 		});
 
-		DataStream<String> result = CEP.pattern(input, pattern).select(new PatternSelectFunction<Event, String>() {
+		DataStream<String> result = CEP.pattern(input, pattern).flatSelect((p, o) -> {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: We should also have one test with `select` and lambda.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java (hunk 1 above)
Comment (twalthr) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: The input validation caused more errors than it solved. Especially with generic types. For lambdas this validation is limited anyway in a JDK compiler.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java (hunk 2 above)
Comment (twalthr) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Good point. I will remove it.

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java [hunk 5]
```diff
@@ -50,12 +41,12 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.assertTrue;
 
 /**
- * Tests the type extractor for lambda functions.
+ * Tests the type extractor for lambda functions. Many tests only work if the compiler supports
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java]: Some tests, like `flatMapLambda()` are not there anymore? It's because we don't use Tycho anymore but do theses tests now reside somewhere else? If not, we should probably have a new module `flink-tycho-lambda-tests` where we test those things that only work with Tycho. Otherwise we lose that coverage.

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java [hunk 6]
```diff
@@ -271,19 +214,20 @@ public void testKeySelectorLambda() {
...
 		Partitioner<Tuple2<Integer, String>> partitioner = (key, numPartitions) -> key.f1.length() % numPartitions;
-		final TypeInformation<?> ti = TypeExtractor.getPartitionerTypes(partitioner);
-
-		Assert.assertTrue(ti.isTupleType());
-		Assert.assertEquals(2, ti.getArity());
-		Assert.assertEquals(((TupleTypeInfo<?>) ti).getTypeAt(0), BasicTypeInfo.INT_TYPE_INFO);
-		Assert.assertEquals(((TupleTypeInfo<?>) ti).getTypeAt(1), BasicTypeInfo.STRING_TYPE_INFO);
+		final TypeInformation<?> ti = TypeExtractor.getPartitionerTypes(partitioner, null, true);
 
+		if (!(ti instanceof MissingTypeInfo)) {
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java]: Why did you need to add this?

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/JarFileCreatorTest.java [hunk 7]
```diff
@@ -102,15 +104,62 @@ public void TestAnonymousInnerClassTrick4() throws Exception {
...
 
 		Assert.assertTrue("Jar file for Anonymous Inner Class is not correct", validate(ans, out));
 
-		out.delete();
+		Assert.assertTrue(out.delete());
+	}
+
+	@Ignore // this is currently not supported (see FLINK-9520)
+	@Test
+	public void testFilterWithMethodReference() throws Exception {
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/JarFileCreatorTest.java]: Did these also not work when they were still in the java8 module?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 8]
```diff
@@ -237,19 +248,81 @@ private boolean validateKeyTypeIsHashable(TypeInformation<?> type) {
...
+	/**
+	 * Adds a type information hint about the key type of a key selector. This method
+	 * can be used in cases where Flink cannot determine automatically what the produced
+	 * type of a key selector is. That can be the case if the selector uses generic type variables
+	 * in the return type that cannot be inferred from the input type.
+	 *
+	 * @param typeInfo type information as a key type hint
+	 * @return This operator with a given key type hint.
+	 */
+	public KeyedStream<T, KEY> returns(TypeInformation<KEY> typeInfo) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: Should probably make this `@PublicEvolving` for now.

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java (hunk 5 above)
Comment (twalthr) [reply flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java]: A flat map can not work by definition because it contains generics. I will add some tests for the `returns`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/JarFileCreatorTest.java (hunk 7 above)
Comment (twalthr) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/JarFileCreatorTest.java]: No this has never worked. I just noticed this bug when I looked at the test. What is this `JarFileCreator` good for actually?

Comment (aljoscha) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/JarFileCreatorTest.java]: It was a util for programmatically creating Jar files. I think it's only used in tests right now, for example `ClassLoaderUtilsTest`.

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java (hunk 6 above)
Comment (twalthr) [reply flink-core/src/test/java/org/apache/flink/api/java/typeutils/LambdaExtractionTest.java]: In case a compiler adds generics, this case would be activated. Otherwise testing for `MissingTypInfo` is correct.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 9]
```diff
@@ -545,15 +543,15 @@ public IntervalJoined(
...
-				TypeExtractor.NO_INDEX,         // output arg indices
-				left.getType(),                 // input 1 type information
-				right.getType(),                // input 2 type information
-				INTERVAL_JOIN_FUNC_NAME ,
-				false
+				TypeExtractor.NO_INDEX,
+				left.getType(),
+				right.getType(),
+				Utils.getCallLocationName(),
+				true
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: Just saw that this should be false. Will correct that.

**Comment by twalthr** [comment]
Thank you @aljoscha. I will add a test for `flapMap` and the bug that I just noticed and merge this.
